main.go

```diff
@@ -0,0 +1,177 @@
+// This file creates an HTTP+Websockets server that implements an in-memory, realtime, bi-directional PubSub system.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"github.com/gorilla/websocket"
+	"github.com/satori/uuid"
+)
+
+// Define an upgrader to upgrade the basic HTTP connection to a websocket
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+	CheckOrigin:     func(r *http.Request) bool { return true },
+}
+
+// The Pub-Sub system maintains a list of subscribers (clients)
+type PubSub struct {
+	Clients []Client
+}
+
+// Every client maintains a unique ID and a websocket connection with the server  
+type Client struct {
+	Id         string
+	Connection *websocket.Conn
+}
+
+// The message consists of action, topic and message fields. The topic functionality can be added to the Pub-Sub system using this struct.
+type Message struct {
+	Action  string          `json:"action"`
+	Topic   string          `json:"topic"`
+	Message json.RawMessage `json:"message"`
+}
+
+// Function to generate a unique ID for every client.
+// Returns:
+// string - A unique identifier string.
+func autoId() string {
+	return uuid.Must(uuid.NewV4(), nil).String()
+}
+
+var ps = &PubSub{}
+
+// Function to set up a basic HTTP server that listens on port 8080
+// and upgrade incoming WebSocket connections. It handles WebSocket 
+// connection requests and upgrades them using the Upgrader method.
+// Parameters:
+// w: http.ResponseWriter - The response writer to write HTTP responses.
+// r: *http.Request - The incoming HTTP request.
+func webSocketHandler(w http.ResponseWriter, r *http.Request) {
+	//Upgrade this connection to a WebSocket connection
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Create a client and assign it a Unique ID
+	client := Client{
+		Id:         autoId(),
+		Connection: ws,
+	}
+
+	// Send a message to the client
+	fmt.Printf("Client Connected:%s", client.Id)
+	err = ws.WriteMessage(1, []byte("Hi Client!"))
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Add client to the list of clients
+	ps.AddClient(client)
```

Comment:
There's a potential race condition here when multiple clients connect simultaneously. You may need to consider adding a `sync.Mutex` or similar to protect `ps.Clients`.